### PR TITLE
fix: Add domain to build_client

### LINKED ISSUE #108: Client always initializes with NoOpProvider
Hi, attempting to follow the README instructions always results in initializing with the NoOp provider, even though it should be the InMemoryProvider. Attempting to use other Providers causes similar errors.

The following is the entirety of my Ruby program:

```ruby
# frozen_string_literal: true

require "open_feature/sdk"

OpenFeature::SDK.configure do |config|
  config.set_provider(OpenFeature::SDK::Provider::InMemoryProvider.new(
    {
      "flag1" => true,
      "flag2" => 1
    }
  ))
end

# Create a client
client = OpenFeature::SDK.build_client(name: "my-app")
pp client
```

Anyone know what might be going on?

### PR DESCRIPTION
**Pull request by grepfruit19**
## This PR

- Adds domain lookup to build_client

### Related Issues

Fixes #108 

### How to test

Instantiate the library with and without a `domain` for its provider.

### CODE REVIEW

File: lib/open_feature/sdk/api.rb [hunk 1]
```diff
@@ -42,10 +42,12 @@ def configure(&block)
         block.call(configuration)
       end
 
-      def build_client(name: nil, version: nil)
+      def build_client(name: nil, version: nil, domain: nil)
         client_options = Metadata.new(name: name, version: version).freeze
```
Comment (beeme1mr) [inline lib/open_feature/sdk/api.rb:46]: Could you please also add domain to the metadata?

**Review by beeme1mr — COMMENTED** [review]
Could you please add some tests and update the readme?

@maxveldink, how would you feel about removing the name argument entirely? It would be a breaking change but help us avoid technical debut as we move towards a 1.0 release.

**Comment by grepfruit19** [comment]
@beeme1mr @maxveldink 

I added some tests.

I didn't want to change the behavior of the library too much, but maybe trying to instantiate a client against a domain that doesn't exist should throw or something, I feel like instantiating with the NoOp provider is unexpected behavior.

**Comment by maxveldink** [comment]
Thanks for fixing this @grepfruit19! Definitely, an oversight I had when I was trying to convert us over from `name` to `domain.` I agree that we should completely drop `name` wherever it's used (pre-1.0, I don't think we should have the idea of "legacy" fields; we should attempt to match the spec as close as possible).

Further, I think dropping `version` makes sense as well. I think @technicalpickles or @josecolella could explain why it was added, but it's not a part of the current specification and might confuse users. One other thing to think through; I don't think we can reuse `Metadata` for both `Provider`s and `Client`s; they have different required fields. I'd be up for either dropping that class in favor of a simple hash for now or introducing `ClientMetadata` and `ProviderMetadata` classes to explicitly state the keys we allow.

Let me know if that's enough to go on, or if you need more direction here. This will definitely be a breaking change.

**Comment by grepfruit19** [comment]
Sounds good, I can add the changes to drop `name` and `version if we're in agreement. I'd rather keep this PR limited in scope as far as adding the `ClientMetadata` and `ProviderMetadata` classes however, so I'd prefer not to add those changes in this PR.

**Comment by grepfruit19** [comment]
@beeme1mr @maxveldink 

I've removed `name` from the library, I think things should be good to go now. I've held off on removing `version` so that y'all can make a separate PR since it seems like that has some further discussion to be had.

Let me know if there's anything else to be done!

File: README.md [hunk 2]
```diff
@@ -54,7 +53,9 @@ OpenFeature::SDK.configure do |config|
 end
 
 # Create a client
-client = OpenFeature::SDK.build_client(name: "my-app")
+client = OpenFeature::SDK.build_client()
```
Comment (maxveldink) [inline README.md:56]: suggestion(non-blocking):

```suggestion
client = OpenFeature::SDK.build_client
```

File: lib/open_feature/sdk/api.rb [hunk 3]
```diff
@@ -42,10 +42,12 @@ def configure(&block)
...
-      def build_client(name: nil, version: nil)
-        client_options = Metadata.new(name: name, version: version).freeze
-        provider = Provider::NoOpProvider.new if provider.nil?
-        Client.new(provider: provider, client_options: client_options, context: context)
+      def build_client(version: nil, domain: nil)
+        client_options = Metadata.new(version: version, domain: domain).freeze
+
+        active_provider = provider(domain:).nil? ? Provider::NoOpProvider.new : provider(domain:)
+
+        Client.new(provider: active_provider, client_options: client_options, context: context)
```
Comment (maxveldink) [inline lib/open_feature/sdk/api.rb:50]: suggestion(non-blocking): If we use hash omission, we should be consistent and do it everywhere. 

```suggestion
        client_options = Metadata.new(version, domain:).freeze

        active_provider = provider(domain:).nil? ? Provider::NoOpProvider.new : provider(domain:)

        Client.new(provider: active_provider, client_options:, context:)
```

File: lib/open_feature/sdk/metadata.rb
Comment (maxveldink) [inline lib/open_feature/sdk/metadata.rb:1]: Confusingly, the `name` here actually refers to the [Provider name](https://openfeature.dev/specification/sections/providers#requirement-211) which is _not_ the same as the client `name` that was replaced by `domain`. We don't need to change anything about metadata and the default provider naming. Sorry if that was unclear 😞

File: lib/open_feature/sdk/client.rb [hunk 4]
```diff
@@ -8,7 +8,7 @@ class Client
       RESULT_TYPE = %i[boolean string number object].freeze
       SUFFIXES = %i[value details].freeze
 
-      attr_reader :metadata
+      attr_reader :metadata, :provider
```
[the remark this answers is not in the document]
Comment (maxveldink) [reply lib/open_feature/sdk/client.rb:11]: I _think_ having read access to the provider is fine here. I want to double-check with someone else if they think this is potentially circumventing the spec.

cc @josecolella @technicalpickles @beeme1mr

**Review by maxveldink — CHANGES_REQUESTED** [review]
We need to back out the changes to the `Metadata#name` field and keep that the same. The changes to `build_client` are good though 👍🏻

**Review by maxveldink — DISMISSED** [review]
Going to wait a little while longer for others to review as well, will merge in by EoD

File: lib/open_feature/sdk/api.rb [hunk 5]
```diff
@@ -42,10 +42,12 @@ def configure(&block)
...
       end
 
-      def build_client(name: nil, version: nil)
-        client_options = Metadata.new(name: name, version: version).freeze
-        provider = Provider::NoOpProvider.new if provider.nil?
-        Client.new(provider: provider, client_options: client_options, context: context)
+      def build_client(name: nil, version: nil, domain: nil)
+        client_options = Metadata.new(name: name, version: version, domain: domain).freeze
+
+        active_provider = provider(domain:).nil? ? Provider::NoOpProvider.new : provider(domain:)
```
Comment (beeme1mr) [inline lib/open_feature/sdk/api.rb:48]: This is fine for now but eventually, the Ruby SDK should allow for providers to be bound to existing clients. That's how other SDKs behave but definitely adds to the complexity.

File: lib/open_feature/sdk/client.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (beeme1mr) [reply lib/open_feature/sdk/client.rb:11]: The provider is not intended to be used directly by a user. I would avoid exposing it here if possible. What use cases are you trying to support?

FYI @toddbaert.

**Review by beeme1mr — DISMISSED** [review]
Approved, but please double-check if exposing the provider is required before merging.

**Comment by grepfruit19** [comment]
@beeme1mr Is there a good way to check what provider a client is using then?

**Comment by beeme1mr** [comment]
> @beeme1mr Is there a good way to check what provider a client is using then?

Hey @grepfruit19, you can use the provider metadata on the client. [Here's](https://github.com/open-feature/js-sdk/blob/main/packages/server/src/client/open-feature-client.ts#L62) how we implemented that in the JS SDK.

If you're interested, this is how we passed the [provider reference](https://github.com/open-feature/js-sdk/blob/main/packages/server/src/client/open-feature-client.ts#L44-L46) to the client.

**Comment by grepfruit19** [comment]
@beeme1mr Then in terms of this ticket, would you be okay with me un-exposing the provider and holding off on exposing the provider metadata for future work?

**Comment by beeme1mr** [comment]
> @beeme1mr Then in terms of this ticket, would you be okay with me un-exposing the provider and holding off on exposing the provider metadata for future work?

Sure, no problem at all. Thanks for your work on this PR.

**Comment by grepfruit19** [comment]
@beeme1mr I've updated the PR and commented out the corresponding tests

**Comment by grepfruit19** [comment]
@beeme1mr I think @maxveldink is out of office today, if this looks good to you mind giving it an approve + merge?

**Review by beeme1mr — APPROVED** [review]

**Pull request merged**